application/src/main/kotlin/net/corda/v5/application/crypto/SigningService.kt

```diff
@@ -20,10 +21,36 @@ import java.util.*
 @DoNotImplement
 interface SigningService {
 
+    /**
+     * Using the provided signing [PublicKey], internally chooses the most appropriate signing spec,
+     * looks up the matching [PrivateKey] and signs the data, selecting the most appropriate signing type.
+     * This overload is preferred; see the other overload if you need to control the signing type.
+     *
+     * @param payload The data to sign over using the chosen key.
+     * @param metadata Optional map of string configuration metadata to be included in the signature. If the metadata
+     *                 is null, or if there are no know entries in the map, that's equivalent to calling `sign` without
+     *                 metadata.
+     * @param publicKey The [PublicKey] partner to an internally held [PrivateKey], either derived from the node's
+     * primary identity, or previously generated via the freshKey method. If the [PublicKey] is actually
+     * a [CompositeKey], the first leaf signing key hosted by the node is used.
+     * @param digestAlgorithmName the message digest algorithm name; if null have the crypto service choose one.
+     * @return A [DigitalSignature.WithKey] representing the signed data and the [PublicKey] that belongs to the
+     * same [KeyPair] as the [PrivateKey] that signed the data. The signature type will be determined automatically.
+     *
+     * @throws [CordaRuntimeException] If the input key is not a member of [keys].
+     */
+    @Suspendable
+    fun sign(payload: ByteArray, metadata: Map<String, String>?, publicKey: PublicKey, digestAlgorithmName: DigestAlgorithmName?): DigitalSignature.WithMetadata
```

Comment:
In practice we will tend to be signing over a SecureHash from some merkle tree. Should we provide a version that takes SecureHash for payload? Perhaps that should be the only version? We would have to decide if we can avoid a second round of hashing if the input is of same type as the digest and there is a compatible signaturespec. I think we also need to document precisely how we form the bytes that we sign at the raw level for interop purposes.